network/p2p/ping.go

```diff
@@ -0,0 +1,184 @@
+type PingService struct {
+	host           host.Host
+	pingProtocolID protocol.ID
+	version        string
+	logger         zerolog.Logger
+}
+
+func NewPingService(h host.Host, pingProtocolID protocol.ID, version string, logger zerolog.Logger) *PingService {
+	ps := &PingService{host: h, pingProtocolID: pingProtocolID, version: version, logger: logger}
+	h.SetStreamHandler(pingProtocolID, ps.PingHandler)
+	return ps
+}
+
+// PingHandler receives the inbound stream for Flow ping protocol and respond back with the PingResponse message
+func (ps *PingService) PingHandler(s network.Stream) {
+
+	errCh := make(chan error, 1)
+	defer close(errCh)
+	timer := time.NewTimer(pingTimeout)
+	defer timer.Stop()
+
+	go func() {
+		log := ps.streamLogger(s)
+		select {
+		case <-timer.C:
+			// if read or write took longer than configured timeout, then reset the stream
+			log.Error().Msg("ping timeout")
+			err := s.Reset()
+			if err != nil {
+				log.Error().Err(err).Msg("failed to reset stream")
+			}
+		case err, ok := <-errCh:
+			// if and error occur while responding to a ping request, then reset the stream
+			if ok {
+				log.Error().Err(err).Msg("failed to ping remote node")
+				err := s.Reset()
+				if err != nil {
+					log.Error().Err(err).Msg("failed to reset stream")
+				}
+				return
+			}
+			// if no error, then just close the stream
+			err = s.Close()
+			if err != nil {
+				log.Error().Err(err).Msg("failed to close stream")
+			}
+		}
+	}()
+
+	pingRequest := &message.PingRequest{}
+	// create the reader
+	reader := ggio.NewDelimitedReader(s, maxPingMessageSize)
+
+	// read the request
+	err := reader.ReadMsg(pingRequest)
+	if err != nil {
+		errCh <- err
+		return
+	}
+
+	// create the writer
+	bufw := bufio.NewWriter(s)
+	writer := ggio.NewDelimitedWriter(bufw)
+
+	// create a PingResponse
+	pingResponse := &message.PingResponse{
+		Version: ps.version, // the semantic version of the build
+		// TODO populate BlockHeight:
+	}
+
+	// send the PingResponse
+	err = writer.WriteMsg(pingResponse)
+	if err != nil {
+		errCh <- err
+		return
+	}
+
+	// flush the stream
+	err = bufw.Flush()
+	if err != nil {
+		errCh <- err
+		return
+	}
+}
+
+// Ping sends a Ping request to the remote node and returns the response, rtt and error if any.
+func (ps *PingService) Ping(ctx context.Context, p peer.ID) (message.PingResponse, time.Duration, error) {
+
+	// create a done channel to indicate Ping request-response is done or an error has occurred
+	done := make(chan error, 1)
+	defer close(done)
+
+	// create a new stream to the remote node
+	s, err := ps.host.NewStream(ctx, p, ps.pingProtocolID)
+	if err != nil {
+		return message.PingResponse{}, -1, err
+	}
+
+	// reset stream on completion
+	go func() {
+		log := ps.streamLogger(s)
+		select {
+		case <-ctx.Done():
+			// context timed out, log an error
+			log.Error().Msg("context timed out on ping to remote node")
+		case <-done:
+		}
+		err := s.Reset()
```

Comment:
A few questions about this: 

- Assume that our context times out and we reach here. Then does it mean that we have only _one single shot_ to reset the stream? why only a single shot?
- Why we do reset the stream upon completion of ping?